crates/unleash-edge/src/main.rs

```diff
@@ -1,13 +1,35 @@
 use clap::Parser;
+use futures::future::join_all;
 use std::net::SocketAddr;
+use std::pin::pin;
+use tokio::signal;
+use tokio::signal::unix::{SignalKind, signal};
 use tower_http::normalize_path::NormalizePathLayer;
 use tracing::info;
 use tracing_subscriber::layer::SubscriberExt;
 use tracing_subscriber::util::SubscriberInitExt;
 use unleash_edge::configure_server;
 use unleash_edge_cli::{CliArgs, EdgeMode};
-use unleash_edge_types::EdgeResult;
 use unleash_edge_types::errors::EdgeError;
+use unleash_edge_types::{BackgroundTask, EdgeResult};
+
+async fn shutdown_signal(address: String, shutdown_tasks: Vec<BackgroundTask>) {
```

Comment:
The shutdown hook is immediately executed by Axum and when it completes Axum is shutdown. So we're responsible for doing the waiting here. The logic here is just to wait for SIGINT (ctrl+c) or SIGTERM (system requests shutdown). Once we get the appropriate signal we execute the shutdown tasks in parallel (they were previously executed in sequence). The exception to this is the metrics tasks which will be sequential related to one another but parallel related to the other background tasks.